traffic_incident_parser/src/traffic_incident_parser_worker.cpp

```diff
@@ -0,0 +1,169 @@
+        mobility_strategy_params.erase(0, pos + delimiter.length());
+    }
+    vec.push_back(mobility_strategy_params);
+
+    std::string lat=vec[0];
+    std::string lon=vec[1];
+    std::string closed_lanes=vec[2];
+    std::string downtrack=vec[3];
+    std::string uptrack=vec[4];
+    latitude=stod(stringParserHelper(lat,lat.find_last_of("lat:")));
+    longitude=stod(stringParserHelper(lon,lon.find_last_of("lon:")));
+    closed_lane=stoi(stringParserHelper(closed_lanes,closed_lanes.find_last_of("lat:")));
+    down_track=stod(stringParserHelper(downtrack,downtrack.find_last_of("lon:")));
+    up_track=stod(stringParserHelper(uptrack,uptrack.find_last_of("lon:")));
+  }
+
+  std::string TrafficIncidentParserWorker::stringParserHelper(std::string str,int str_index)
+  {
+    std::string str_temp="";
+    for(int i=str_index+1;i<=str.length();i++)
+    {
+        str_temp+=str[i];
+    }
+    return str_temp;
+  }
+
+ lanelet::BasicPoint2d TrafficIncidentParserWorker::earthToMapFrame()
+  {
+    lanelet::projection::LocalFrameProjector projector(projection_msg_.c_str());
+    lanelet::GPSPoint gps_point;
+    gps_point.lat = latitude;
+    gps_point.lon = longitude;
+    gps_point.ele = 0;
+    auto local_point3d = projector.forward(gps_point);
+    return {local_point3d.x(), local_point3d.y()};
+  }
+
+  cav_msgs::TrafficControlMessageV01 TrafficIncidentParserWorker::composeTrafficControlMesssage()
+  {
+    local_point_=earthToMapFrame();
+    auto current_lanelets = lanelet::geometry::findNearest(wm_->getMap()->laneletLayer, local_point_, 1);  
+    lanelet::ConstLanelet current_lanelet = current_lanelets[0].second;
+    double following_distance=0.0;
+    double previous_distance=0.0;
+    std::vector<lanelet::BasicPoint2d> center_line_points_right={};
+    std::vector<lanelet::BasicPoint2d> center_line_points_left={};
+ 
+    while(following_distance<down_track)
+    {
+      auto next_lanelets = wm_->getMapRoutingGraph()->following(current_lanelet);
+      if(!next_lanelets.empty())
+      {
+        carma_wm::TrackPos tp = carma_wm::geometry::trackPos(next_lanelets[0], local_point_);
+        following_distance= std::fabs(tp.downtrack);
+      }
+      
+      auto back_point = current_lanelet.centerline().back().basicPoint2d();
+      auto front_point = current_lanelet.centerline().front().basicPoint2d();
+      lanelet::BasicPoint2d middle = {(back_point.x() + front_point.x())/2, (back_point.y() + front_point.y())/2};
+          
+      center_line_points_right.push_back(middle);
+      if(next_lanelets.empty())
+      {
+        break;
+      }
+      else
+      {
+      current_lanelet=next_lanelets[0];
+      }
+    }
+
+    current_lanelets = lanelet::geometry::findNearest(wm_->getMap()->laneletLayer, local_point_, 1); 
+    current_lanelet = current_lanelets[0].second;
+    carma_wm::TrackPos tp = carma_wm::geometry::trackPos(current_lanelet, local_point_);
+    previous_distance= std::fabs(tp.downtrack);
+    
```

Comment:
same as above. let's explain the algorithm a little bit. what does this while loop do